src/main/java/com/devonfw/quarkus/productmanagement/domain/repo/ProductFragmentImpl.java

```diff
@@ -28,87 +25,33 @@
-    Root<ProductEntity> root = cq.from(ProductEntity.class);
-    List<Predicate> predicates = new ArrayList<>();
-    CriteriaBuilder cb = this.em.getCriteriaBuilder();
-    if (dto.getTitle() != null && !dto.getTitle().isEmpty()) {
-      predicates.add(cb.like(root.get(ProductEntity_.TITLE), dto.getTitle()));
-    }
-
-    if (dto.getPrice() != null) {
-      predicates.add(cb.gt(root.get(ProductEntity_.PRICE), dto.getPrice()));
-    } else if (dto.getPriceMin() != null || dto.getPriceMax() != null) {
-      if (dto.getPriceMin() != null) {
-        predicates.add(cb.gt(root.get(ProductEntity_.PRICE), dto.getPriceMin()));
-      }
-      if (dto.getPriceMax() != null) {
-        predicates.add(cb.lt(root.get(ProductEntity_.PRICE), dto.getPriceMax()));
-      }
-    }
-
-    if (!predicates.isEmpty())
-
-    {
-      cq.where(predicates.toArray(new Predicate[0]));
-    }
-
-    // Order by title
-    cq.orderBy(cb.desc(root.get(ProductEntity_.TITLE)));
-
-    TypedQuery<ProductEntity> products = this.em.createQuery(cq).setFirstResult(dto.getPageNumber() * dto.getPageSize())
-        .setMaxResults(dto.getPageSize());
-    return new PageImpl<ProductEntity>(products.getResultList(), PageRequest.of(dto.getPageNumber(), dto.getPageSize()),
-        products.getResultList().size());
-  }
-
-  @Override
-  public Page<ProductEntity> findAllQueryDsl(ProductSearchCriteriaDto dto) {
+  public Page<ProductEntity> findByCriteria(ProductSearchCriteriaDto searchCriteria) {
 
     QProductEntity product = QProductEntity.productEntity;
-    JPAQuery<ProductEntity> query = new JPAQuery<ProductEntity>(this.em);
-    query.from(product);
-    if (dto.getTitle() != null && !dto.getTitle().isEmpty()) {
-      query.where(product.title.eq(dto.getTitle()));
+    List<Predicate> predicates = new ArrayList<>();
+    if (!isEmpty(searchCriteria.getTitle())) {
+      predicates.add(product.title.eq(searchCriteria.getTitle()));
     }
-
-    if (dto.getPrice() != null) {
-      query.where(product.price.eq(dto.getPrice()));
-    } else if (dto.getPriceMin() != null || dto.getPriceMax() != null) {
-      if (dto.getPriceMin() != null) {
-        query.where(product.price.gt(dto.getPriceMin()));
+    if (!isNull(searchCriteria.getPrice())) {
+      predicates.add(product.price.eq(searchCriteria.getPrice()));
+    } else {
+      if (!isNull(searchCriteria.getPriceMin())) {
+        predicates.add(product.price.gt(searchCriteria.getPriceMin()));
       }
-      if (dto.getPriceMax() != null) {
-        query.where(product.price.lt(dto.getPriceMax()));
+      if (!isNull(searchCriteria.getPriceMax())) {
+        predicates.add(product.price.lt(searchCriteria.getPriceMax()));
       }
     }
 
-    // Order by title
+    JPAQuery<ProductEntity> query = new JPAQuery<ProductEntity>(this.em).from(product);
+    query.where(predicates.toArray(Predicate[]::new));
     query.orderBy(product.title.desc());
+    long total = query.fetchCount();
```

Comment:
Added determineTotal in searchCriteria. This way the UI will have a control over total. WDYT? 
```
public class ProductSearchCriteriaDto {
  .........................
  @QueryParam("determineTotal")
  private Boolean determineTotal;
}
```

Reply:
Not sure how `Pageable `get automatically injected  in (like in springboot application) to the SearchCriteria. Does the following way look good?

```
public abstract class ApplicationSearchCriteriaDto {
  private int pageNumber = 0;
  private int pageSize = 10;
  private Pageable pageable;
  
  public Pageable getPageable() {
    return PageRequest.of(pageNumber, pageSize);
  }
}
```

Reply:
My idea was rather to implement `Pageable` or extend the base class directly in ApplicationSearchCriteriaDto,
or to aggregate it instead of using own properties like `pageNumber` and `pageSize`.

Reply:
IMHO, it can be done in following ways.

1. Pageable can be implemented, but there are many methods in it, we may have to implement ourselves. 

2. AbstractPageRequest/PageRequest has can be extended, but those are immutable classes, we may have to define page and size variables and need to call its constructor. 

3. Another way is to implement pageable and make ApplicationSearchCriteriaDto as a proxy class of PageRequest, since it is immutable.

Let me know your thoughts.

Reply:
Maybe the easiest approach would be to add a `asPageable` Method to `ApplicationSearchCriteriaDto` that creates a `Pageable` from the single attributes. This also prevents to tight coupling with Spring-Data that could impact JSON structure changes in case of spring internal changes in the future.